DRAFT: q-implant-qparam-test (#11579)
Samsung/ONE: compiler/q-implant/README.md

# q-implant

> _q-implant_ writes quantization parameters and weights (given as .json and .npy files) to a circle model.

Reviewer: ```suggestion
_q-implant_ embeds quantization parameters and weights (given as .json and .npy files) into circle model.
```

Reviewer: plz explain more about "writes weights (.npy)": where does this .npy file come from?

Reviewer 2: Hello, I'm Junyeong Kang working with @FantBlog. 

Answer 1.

I'm sorry.
This README.md is already merged, but we made a mistake of not matching sync. #11562
We'll post a separate PR to revise README.md by supplementing the feedback section

Answer 2.

> plz explain more about "writes weights (.npy)": where does this .npy file come from?

We create ".npy" files through "[test_utils.py](https://github.com/Samsung/ONE/pull/11579/files/cca7de5fe6654722292d274b5f9c261f2f96769a#diff-c855e669b0b66191648c521df67c731f0b1dc738497894cb0ef5cc5cabc5c3d1R28)". 

".npy" is a numpy-type file used to represent data from quantization parameters and weights in param.json.

We write quantization parameters and weights in the form of ".npy" into the tensor through "q-implant".

Reviewer: `test_utils.py` uses random value generation. we use random value for testing and for testing, `test_utils.py` looks OK.
but for real model quantization, how do we get this npy files?

Reviewer 2: `q-implant` is a module that only writes quantization parameters and weights into the tensor of the existing circle. 
It is correct to use `test_utils.py` to verify this behavior.

Therefore, if users want to use `q-implant` for the real model, they must obtain qparam (param.json + .npy) through the extraction process.

That's why our team tries "q-extract-torch" module.

Reviewer: > Therefore, if users want to use q-implant for the real model, they must obtain qparam (param.json + .npy) through the extraction process.

Then, `q-implant` should provide the structure / specification of this json and npy file.

Reviewer 2: > > Therefore, if users want to use q-implant for the real model, they must obtain qparam (param.json + .npy) through the extraction process.
> 
> 
> 
> Then, `q-implant` should provide the structure / specification of this json and npy file.
> 
> 
> 
> 

As soon as possible, we create pull request about 'q-implant' README.md including previous feedbacks and the structure / specification of json and npy file.

Thank you so much!

Reviewer: I don't think #11638 provides specification for q-implant json file.
Please be more specific.
Update this PR before posting the PR.

Reviewer 2: > I don't think #11638 provides specification for q-implant json file. Please be more specific. Update this Draft PR before posting new PR.

OK. I updated the PR to Draft PR.
I'll update supplementary explanations for q-implant json file soon.

## Example

```sh
q-implant input.circle qparam.json output.circle
```

`qparam.json` and `*.npy` files must exist in the same directory. For example,

```json
{
  "ifm": {
    "dtype": "uint8",
    "scale": "0.npy",

>     "zerop": "1.npy",

Reviewer: above sentence used word `weights` but these don't seem to be weights.

Reviewer: can you add a sample json with input, conv2d(with kernel and bias), output ?

Reviewer 2: > above sentence used word `weights` but these don't seem to be weights.

This README.md was created assuming that "q-implant" was applied to a simple Conv2D model. This model consists of `ifm`, `weight`, and `bias` tensor as input, and `ofm` tensor as output.

In the case of `ifm` written as an example on README.md, it is a tensor that does not include value, so there is no part to write weights, but only quantization parameters.

There seems to be a lack of explanation for this part. We'll add an example for a tensor with value.

Author: >can you add a sample json with input, conv2d(with kernel and bias), output ?

this is sample used in q-implant-op-level-test 

1. input file : Generate circle files from TFLite recipes
"TFLite recipe" -> tflchef -> "tflite file" -> tflite2circle -> "circle file"

2. json file
```
{
  "ifm": {
    "dtype": "uint8",
    "scale": "0.npy",
    "zerop": "1.npy",
    "quantized_dimension": 0
  },
  "ker": {
    "dtype": "uint8",
    "scale": "2.npy",
    "zerop": "3.npy",
    "quantized_dimension": 0,
    "value": "4.npy"
  },
  "bias": {
    "dtype": "int32",
    "scale": "5.npy",
    "zerop": "6.npy",
    "quantized_dimension": 0,
    "value": "7.npy"
  },
  "ofm": {
    "dtype": "uint8",
    "scale": "8.npy",
    "zerop": "9.npy",
    "quantized_dimension": 0
  }
}
```

3. output file : dump to use circledump
```
Dump: /home/ssafy/ONE/ONE/build/compiler/q-implant-op-level-test/Conv2D_000_Q8/output.circle

===================================================================
Model version: 0
 # sub graphs: 1

Operator Codes: [order] OpCodeName (OpCode Enum)
[0] CONV_2D (code: 3, dep_code: 3, version: 1)

Buffers: B(index) (length) values, if any
B(0) (0) 
B(1) (0) 
B(2) (2) 0x6e 0x80
B(3) (4) 0x36 00 00 00
B(4) (0) 

metadata : B(index) name

-------------------------------------------------------------------
Sub-Graph: #0 main

Data Format:
CHANNEL_LAST (NHWC for 2d, NDHWC for 3d data)

Operands: T(subgraph index : tensor index) TYPE (shape) (shape_signature) B(buffer index) (variable) OperandName
T(0:0) UINT8 (1, 3, 3, 2) B(1) ifm
    Quantization: scale(0.0686598) zeropt(169) quantized_dimension(0)

T(0:1) UINT8 (1, 1, 1, 2) B(2) ker
    Quantization: scale(0.582998) zeropt(58) quantized_dimension(0)

T(0:2) INT32 (1) B(3) bias
    Quantization: scale(0.543995) zeropt(164) quantized_dimension(0)

T(0:3) UINT8 (1, 3, 3, 1) B(4) ofm
    Quantization: scale(0.549946) zeropt(35) quantized_dimension(0)

Operators: O(subgraph index : operator index) OpCodeName 
    Option(values) ... <-- depending on OpCode
    I T(tensor index) OperandName <-- as input
    O T(tensor index) OperandName <-- as output
O(0:0) CONV_2D 
    Padding(1) Stride.W(1) Stride.H(1) Dilation.W(1) Dilation.H(1) Activation(NONE)
    I T(0:0) ifm
    I T(0:1) ker
    I T(0:2) bias
    O T(0:3) ofm

Inputs/Outputs: I(input)/O(output) T(tensor index) OperandName
I T(0:0) ifm
O T(0:3) ofm

===================================================================
```

Reviewer: > this is sample used in q-implant-op-level-test

not here ... in to the README.md file.

    "quantized_dimension": 0
  },
  ...
}
```

> `0.npy` and `1.npy` must exist in the same directory with `qparam.json`.

Reviewer: from this explanation, there will be two `.npy` files for one operator(node) from the model. for real models, there could be 10s or 100s for nodes in a model.
which would produces lots of files in the process of `q-implant`.
I don't think this is a good approach for users.

How about using `h5` format and produce single file for a model?

Reviewer 3: > How about using h5 format and produce single file for a model?

There is an issue in using h5 format.

tensors can have `/` in their names, but h5 does not allow `/` in the object name. So we have to replace it into another character, which makes hard to recover the original tensor name.

Reviewer: let's discuss this in 11530 as this file is not related with "operator level test"
